plugins/single_plugins/fisheye.cpp

```diff
@@ -107,110 +108,120 @@ class wayfire_fisheye : public wf::plugin_interface_t
+        OpenGL::render_end();
+    }
 
-        wf::post_hook_t render_hook = [=](const wf::framebuffer_base_t& source,
-            const wf::framebuffer_base_t& dest)
+    wf::activator_callback toggle_cb = [=] (wf::activator_source_t, uint32_t)
+    {
+        if (!output->can_activate_plugin(grab_interface))
         {
-            auto oc = output->get_cursor_position();
-            wlr_box box = {(int)oc.x, (int)oc.y, 1, 1};
-            box = output->render->get_target_framebuffer().
-                framebuffer_box_from_geometry_box(box);
-            oc.x = box.x;
-            oc.y = box.y;
-
-            static const float vertexData[] = {
-                -1.0f, -1.0f,
-                 1.0f, -1.0f,
-                 1.0f,  1.0f,
-                -1.0f,  1.0f
-            };
-
-            OpenGL::render_begin(dest);
-            program.use(wf::TEXTURE_TYPE_RGBA);
-            GL_CALL(glBindTexture(GL_TEXTURE_2D, source.tex));
-            GL_CALL(glActiveTexture(GL_TEXTURE0));
-
-            program.uniform2f("u_mouse", oc.x, oc.y);
-            program.uniform2f("u_resolution", dest.viewport_width, dest.viewport_height);
-            program.uniform1f("u_radius", radius);
-            program.uniform1f("u_zoom", progression);
-
-            program.attrib_pointer("position", 2, 0, vertexData);
-
-            GL_CALL(glDrawArrays (GL_TRIANGLE_FAN, 0, 4));
-            GL_CALL(glBindTexture(GL_TEXTURE_2D, 0));
-
-            program.deactivate();
-            OpenGL::render_end();
-
-            if (!active && !progression.running())
-                finalize();
-        };
+            return false;
+        }
 
-        void finalize()
+        if (active)
         {
-            output->render->rem_post(&render_hook);
-            output->render->set_redraw_always(false);
-            hook_set = false;
+            active = false;
+            progression.animate(0);
+        } else
+        {
+            active = true;
+            progression.animate(zoom);
+            if (!hook_set)
+            {
+                hook_set = true;
+                output->render->add_post(&render_hook);
+                output->render->set_redraw_always();
+            }
         }
 
-        void fini() override
-        {
-            if (hook_set)
-                finalize();
+        return true;
+    };
+
+    wf::post_hook_t render_hook = [=] (const wf::framebuffer_base_t& source,
+                                       const wf::framebuffer_base_t& dest)
+    {
+        auto oc     = output->get_cursor_position();
+        wlr_box box = {(int)oc.x, (int)oc.y, 1, 1};
+        box = output->render->get_target_framebuffer().
+            framebuffer_box_from_geometry_box(box);
+        oc.x = box.x;
+        oc.y = box.y;
+
+        static const float vertexData[] = {
+            -1.0f, -1.0f,
+            1.0f, -1.0f,
+            1.0f, 1.0f,
+            -1.0f, 1.0f
```

Comment:
Is there a way to align the `1`'s with each other?

Reply:
Another good point, I initially didn't consider the alignment options, but there a few of them left which might be useful. I'll experiment and see whether I can get it to work sensibly

Reply:
I couldn't find a way to make this work without breaking a ton of other things ...

Reply:
maybe a bad idea, but for fisheye prefix also with + (never tried this though).